common/http/platform/beast/http_resumer.cpp

```diff
@@ -0,0 +1,454 @@
+			return;
+		}
+
+		auto next_call_err = resumer_next_header_handler.client_.client_.AsyncCall(
+			range_req, resumer_next_header_handler, resumer_next_body_handler);
+		if (next_call_err != error::NoError) {
+			auto err_user = MakeError(
+				DownloadResumerError,
+				"Failed to copy data to schedule the next resumer call: " + next_call_err.String());
+			resumer_next_header_handler.client_.logger_.Error(err_user.String());
+			resumer_next_header_handler.client_.Cancel();
+			resumer_next_header_handler.user_body_handler_(expected::unexpected(err_user));
+		}
+	};
+}
+
+void HeaderHandlerFunctor::operator()(ExpectedIncomingResponsePtr exp_resp) {
+	if (!client_.IsOngoing()) {
+		HandleFirstResponse(exp_resp);
+	} else {
+		HandleNextResponse(exp_resp);
+	}
+
+	if (client_.IsOngoing() && exp_resp) {
+		// Set resumer BodyWriter
+		auto resp = exp_resp.value();
+		auto body_writer =
+			make_shared<io::ByteOffsetWriter>(client_.state_.buffer, client_.state_.buffer->size());
+		body_writer.get()->SetUnlimited(true);
+		resp->SetBodyWriter(body_writer);
+	}
+}
+
+void HeaderHandlerFunctor::HandleFirstResponse(ExpectedIncomingResponsePtr exp_resp) {
+	// The first response shall call the user header callback, both on errors and on success
+	if (!exp_resp) {
+		client_.logger_.Error(exp_resp.error().String());
+		client_.Cancel();
+		user_header_handler_(exp_resp);
+		return;
+	}
+	auto resp = exp_resp.value();
+
+	if (resp->GetStatusCode() != mender::http::StatusOK) {
+		client_.logger_.Error("Unexpected status code " + to_string(resp->GetStatusCode()));
+		client_.Cancel();
+		// Return to the user the incoming response, not an error, so that it can be handled.
+		user_header_handler_(exp_resp);
+		return;
+	}
+
+	client_.state_.init();
+
+	auto exp_header = resp->GetHeader("Content-Length");
+	if (!exp_header || exp_header.value() == "0") {
+		auto err = error::Error(
+			exp_header.error().code, "Response does not contain Content-Length header ");
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		user_header_handler_(expected::unexpected(err));
+		return;
+	}
+
+	auto exp_length = common::StringToLongLong(exp_header.value());
+	if (!exp_length || exp_length.value() < 0) {
+		auto err = error::Error(
+			exp_length.error().code,
+			"Content-Length contains invalid number: " + exp_header.value());
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		user_header_handler_(expected::unexpected(err));
+		return;
+	}
+
+	// Prepare state and call user handler
+	client_.state_.content_length = exp_length.value();
+	client_.state_.offset = 0;
+	user_header_handler_(exp_resp);
+
+	io::WriterPtr user_writer = resp->GetBodyWriter();
+	if (!user_writer) {
+		auto err = MakeError(
+			DownloadResumerError, "The user did not set a BodyWriter to write the data into");
+		client_.logger_.Error(err.String());
+		client_.Cancel();
+		return;
+	}
+	client_.state_.user_body_writer = user_writer;
```

Comment:
Unfortunately this section will need some care. In hindsight, using a body writer was a mistake (*). Later the `MakeBodyAsyncReader` was added, but I kept `SetBodyWriter` because it is handy for small requests and tests. But I kind of regret it now. Maybe we should discuss this face to face, because I'm not completely sure what's the best solution here, and I'm also deep into http readers and writers as part of the http forwarding.

(*) First of all, [error reporting is not possible](https://github.com/mendersoftware/mender/commit/5251a1a6fdb6051b2ddb5958f89019a8c1fc1db8), and we also need to make it an async interface in order to be able to multitask operations, which we use for many things, dbus, update module process interaction, soon http forwarding, etc.

Reply:
As a primer for this discussion, I just opened #1371 which finalizes the async support. Perhaps you want to review / learn that one for more context, and we can take the discussion afterwards.

Reply:
Face to face on Monday, please :)

Reply:
As discussed on Monday, I will keep the PR on hold until you submit some rework on the `http::Client` and I can adapt my code to address this point.